### ISSUE TITLE
Implement lexer on a stream

### ISSUE THREAD

**Issue by xrchz** [CakeML/cakeml#645]
This issue is to make the lexer take an input stream, and add some facility for turning strings into input streams. This probably subsumes #353 although is probably harder.

**Comment by mn200** [comment]
The followup would be to have the parser do the same. 

It seems as if the logic could model the stream as a lazy list, but have compilation magically treat that as something of type `unit -> thing`.

**Comment by mn200** [comment]
Certainly, the `LUNFOLD` function connects a state-monad function to a lazy list...
